packages/ckeditor5-markdown-gfm/src/pastefrommarkdownexperimental.ts

```diff
@@ -0,0 +1,195 @@
+ */
+
+import { type Editor, Plugin } from 'ckeditor5/src/core';
+import { Clipboard, type ClipboardInputTransformationEvent, type ClipboardPipeline } from 'ckeditor5/src/clipboard';
+import GFMDataProcessor from './gfmdataprocessor';
+import type { ViewDocumentKeyDownEvent } from 'ckeditor5/src/engine';
+
+/**
+ * The GitHub Flavored Markdown (GFM) paste plugin.
+ *
+ * For a detailed overview, check the {@glink features/pasting/paste-markdown Markdown feature} guide.
+ */
+export default class PasteFromMarkdownExperimental extends Plugin {
+	/**
+	 * @internal
+	 */
+	private _gfmDataProcessor: GFMDataProcessor;
+
+	/**
+	 * @inheritDoc
+	 */
+	constructor( editor: Editor ) {
+		super( editor );
+
+		this._gfmDataProcessor = new GFMDataProcessor( editor.data.viewDocument );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public static get pluginName() {
+		return 'PasteFromMarkdownExperimental' as const;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public static get requires() {
+		return [ Clipboard ] as const;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public init(): void {
+		const editor = this.editor;
+		const view = editor.editing.view;
+		const viewDocument = view.document;
+
+		const clipboardPipeline: ClipboardPipeline = editor.plugins.get( 'ClipboardPipeline' );
+
+		let shiftPressed = false;
+
+		this.listenTo<ViewDocumentKeyDownEvent>( viewDocument, 'keydown', ( evt, data ) => {
+			shiftPressed = data.shiftKey;
+		} );
+
+		this.listenTo<ClipboardInputTransformationEvent>( clipboardPipeline, 'inputTransformation', ( evt, data ) => {
+			if ( shiftPressed ) {
+				return;
+			}
+
+			const dataAsTextHtml = data.dataTransfer.getData( 'text/html' );
+
+			if ( !dataAsTextHtml ) {
+				const dataAsTextPlain = data.dataTransfer.getData( 'text/plain' );
+				data.content = this._gfmDataProcessor.toView( dataAsTextPlain );
+
+				return;
+			}
+
+			const markdownFromHtml = this._parseMarkdownFromHtml( dataAsTextHtml );
+
+			if ( markdownFromHtml ) {
+				data.content = this._gfmDataProcessor.toView( markdownFromHtml );
+			}
+		} );
+	}
+
+	/**
+	 * Determines if code copied from a website in `text/html` type can be parsed as markdown.
+	 * It removes any OS specific HTML tags e.g. <meta> on Mac OS and <!--StartFragment--> on Windows.
+	 * Then removes a single wrapper HTML tag or wrappers for sibling tags, and if there are no more tags left,
+	 * returns the remaining text. Returns null, if there are any remaining HTML tags detected.
+	 *
+	 * @private
+	 * @param htmlString Clipboard content in `text/html` type format.
+	 */
+	private _parseMarkdownFromHtml( htmlString: string ): string | null {
+		const withoutOsSpecificTags = this._removeOsSpecificTags( htmlString );
+
+		const withoutWrapperTag = this._isHtmlList( withoutOsSpecificTags ) ?
+			this._removeListWrapperTagsAndBrs( withoutOsSpecificTags ) :
+			this._removeWrapperTag( withoutOsSpecificTags );
+
+		if ( this._containsAnyRemainingHtmlTags( withoutWrapperTag ) ) {
```

Comment:
Valid Markdown can contain HTML tags (for instance, complicated tables). It seems like we don't support them here. Is it intentional?

Reply:
We had a lot of discussions about this topic, and our current goal is to provide full paste support in format `text/plain` - code copied from online markdown editors and code blocks. We decided to provide limited support for pasting code in format `text/html` - copied from a website but not in a code block.

We added support for simple html structures, so pasting most code from here should work fine: https://www.markdownguide.org/basic-syntax/

Unfortunately tables are a bit more complex, and the algorithm detecting them would be much more complicated.

We tried to follow how Notion handles pasting markdown, and it behaves almost exactly the same as our implementation - including tables.

Reply:
Plus, this plugin is released as experimental. If there are any requests from the community for paste markdown table from HTML website support, we can get back to it then.